### PR TITLE
Use `$/logTrace` for server trace logs in Zed and VS Code

### PR DESCRIPTION
**Pull request by osiewicz**
## Summary

This pull request adds support for logging via `$/logTrace` RPC messages. It also enables that code path for when a client is Zed editor (as there's no way for us to generically tell whether a client prefers `$/logTrace` over stderr.

Related to: #12523
/cc @dhruvmanila 

## Test Plan

I've built Ruff from this branch and tested it manually with Zed.

### CODE REVIEW

File: crates/ruff_server/src/trace.rs [hunk 1]
```diff
@@ -82,15 +121,25 @@ pub(crate) fn init_tracing(
                 .ok()
         });
 
+    let logger = match log_file {
+        Some(file) => BoxMakeWriter::new(Arc::new(file)),
+        None => {
+            if client
+                .as_ref()
+                .is_some_and(|client| client.name.starts_with("Zed ") || client.name == "Zed")
```
Comment (dhruvmanila) [inline crates/ruff_server/src/trace.rs:129]: nit: can we combine both `starts_with` and `==` into a single check that just uses `starts_with("Zed")`? I think that should work as per the possible display names:

https://github.com/zed-industries/zed/blob/b7c6f3e98eb53eb516cba9bb1724ebbbf780cfd4/crates/release_channel/src/lib.rs#L129-L137

File: crates/ruff_server/src/trace.rs [hunk 2]
```diff
@@ -43,13 +49,46 @@ pub(crate) fn set_trace_value(trace_value: TraceValue) {
...
 pub(crate) fn init_tracing(
-    sender: ClientSender,
+    sender: &ClientSender,
     log_level: LogLevel,
     log_file: Option<&std::path::Path>,
+    client: &Option<ClientInfo>,
 ) {
     LOGGING_SENDER
-        .set(sender)
+        .set(sender.clone())
```
Comment (dhruvmanila) [inline crates/ruff_server/src/trace.rs:91]: Any specific reason why `ClientSender` is a reference? We can just pass in the value as it's going to be consumed anyway.

File: crates/ruff_server/src/trace.rs (hunk 1 above)
Comment (dhruvmanila) [reply crates/ruff_server/src/trace.rs:129]: And, I think we should also include "Visual Studio Code" here because it also supports `$/logTrace` requests.

**Review by dhruvmanila — APPROVED** [review]

**Comment by dhruvmanila** [comment]
I pushed both the changes and verified in VS Code:
```
2024-07-30 08:12:09.990 [info] [Trace - 8:12:09 AM]    0.021675875s  INFO ruff:main ruff_server::server: Configuration file watcher successfully registered
```

File: crates/ruff_server/src/trace.rs (hunk 2 above)
Comment (osiewicz) [reply crates/ruff_server/src/trace.rs]: Clippy was just complaining about it in [CI](https://github.com/astral-sh/ruff/actions/runs/10142100234/job/28040714091), no other reason.

File: crates/ruff_server/src/trace.rs (hunk 1 above)
Comment (MichaReiser) [reply crates/ruff_server/src/trace.rs]: Very interesting that there's no `ClientCapability` indicating whether `logTrace` is supported.

**Pull request merged**